### ISSUE TITLE
Go cassandra driver (gocql) connection pool failure after Yugabyte rolling restart 

### ISSUE THREAD

**Issue by justinfx** [yugabyte/yugabyte-db#10182]
(This issue should be checked after the client is updated to the parent via #9818 but may still be an open issue)

Slack discussion: https://yugabyte-db.slack.com/archives/CG0KQF0GG/p1632856106352700

Related parent gocql project issues:
- https://github.com/gocql/gocql/issues/1575
- https://github.com/gocql/gocql/pull/1555
- https://github.com/gocql/gocql/issues/915

In trying out the situation of rolling restarts on the Yugabyte Operator cluster, I have observed a problem where a client application will end up with no hosts left in the connection pool and only a restart of the application will resolve the problem. 

```
gocql: no hosts available in the pool
```

I've done some more testing and what I have found is that using the individual dns names seems to be ideal since they match up with the `broadcast_address` assigned by the operator to each tserver:

```
yb-tserver-0.yb-tservers.my-namespace.svc.cluster.local
yb-tserver-1.yb-tservers.my-namespace.svc.cluster.local
yb-tserver-2.yb-tservers.my-namespace.svc.cluster.local
```

Then if I scale the tservers from 3 -> 2, I will see the connection event logged in my client that one is now timing out (expected). And when I scale the tservers back from 2 -> 3, there is a new ip, but the connection timeout error to the old ip stop logging, and the new one logs as up and things seem happy.
The problem is when the tservers are rolled too quickly, even if I am using `OrderedReady` pod management policy to make them restart one at a time. It seems in this case my client completely loses all hosts and doesn't work until a restart. So I feel like I must be missing some kind of client configuration to cope with this situation. Or maybe the yb operator is rolling them too fast? Or maybe I need to be careful and manually roll tservers. But that doesn't help in the situation where the nodes go down for another reason, and then come back up with new ips, and the application client is broken until restart.

https://github.com/gocql/gocql/issues/1575 refers to a possible issue in the connection pool implementation where hosts are never re-resolved to an IP at a later point in time after the initial connection.

**Comment by ashetkar** [comment]
Thanks.
Will try and verify if this issue still persists once the Yugabyte gocql is synced with the latest upstream gocql.

**Comment by ashetkar** [comment]
Yes, I was using the above code to explicitly resolve dns names to ip addresses.
And by 're-connect' I meant recreating the session (and not restarting the application itself).

Agree that this is something better handled at the driver layer itself rather than in the application.
This issue is still seen despite a fix for https://github.com/gocql/gocql/issues/915 (it is still open).

Looking into the code to better understand it. Will update soon.

**Comment by justinfx** [comment]
Hi @ashetkar. Since this is still an open issue and I need to ensure my application has a stable connection to the cluster, I am going to have to apply the client-side workaround for now. I've been looking at the various Observer hooks in the client and I can see when there are `ObserveConnect` errors with a DOWN state and timeout. But what I am really trying to hook into is the "no hosts available in the pool" error. Do you have advice on the best place to determine that all hosts have been exhausted in the pool so I can trigger a reconnect on the client?  It seems like my two options are:

1. Watch `ObserverConnect` errors for enough DOWN states and then check the host pool manually?
2. Wrap the client to intercept query errors or check every error directly, for `ErrNoConnections`

**Comment by justinfx** [comment]
Thought I would update this ticket with the details about the extent I had to go to, in order to work around this issue on the client application side of things...

Because as the application using the client driver doesn't have enough access to the internal state of the connection pool, I had to do the best I could to use a `ConnectObserver` to watch details about the host events. The high level logic goes something like this:

- Open the client driver, by installing custom [ConnectObserver](https://pkg.go.dev/github.com/yugabyte/gocql#ConnectObserver), and create `Session`. The session access must be guarded by a mutex (`RWMutex`)
- All access to the session must be guarded by a read lock and needs to be "checked out" as a local variable before use. After usage of the session, it has to be checked back in with the lock and determine if it has been replaced in the meantime. If it was replaced, then close the session we checked out.
- `ConnectObserver` maintains a host map count, last timestamp of reconnect, and if it is currently reconnecting. 
  - For every error event that also contains a connect address...
    - if the host is "up", remove it from the map
    - if the host is "down", add it to the map
    - If there is only one configured host and it is down, OR, if more than 2 configured hosts are down
      - check that we have not reconnected in the last 10 seconds
      - check that we are not concurrently reconnecting (concurrent access to client driver)
      - lock-acquire the session, close, and recreate a new session
      - update last reconnect time

The solution is not perfect because it still results in queries hanging a bit while this detection logic plays out. And I had to make sure to handle the fact that I can only "reconnect" by replacing the session, which is in concurrent use. So I have to be careful to guard the access to the session and replace it with a new one without interfering with a query that is already in progress. 

Ideally this would be handled more accurately within the client driver pool and would not need to replace the session object. The driver needs to be smarter in knowing when the host pool is exhausted and needs to re-resolve the originally configured host list dns and update the pool again. It should be as transparent as possible for concurrent consumers of the session.

**Comment by justinfx** [comment]
@ashetkar has there been any progress on this?

**Comment by justinfx** [comment]
@ashetkar this has been reported as fixed in gocql [1.4.0](https://github.com/gocql/gocql/releases) via https://github.com/gocql/gocql/pull/1682

Any chance the yugabyte go client could be updated for these fixes?

**Comment by ashetkar** [comment]
@justinfx Thanks for pointing it out to us. Yes, we need to update the driver to the latest upstream master anyways, because it has quite a sizeable number of commits since we last updated. Let me come back on this in a day or two with an approximate timeline.

**Comment by ashetkar** [comment]
Sorry for the delay in responding. We are targeting this for the next month. Will keep this ticket updated, thanks.

**Comment by ashetkar** [comment]
We have upgraded the YB Gocql driver to the upstream Gocql version 1.6.0 in January this year.
The fix mentioned above in version 1.4.0 of upstream gocql should now be available in this fork as well.
